Initial commit of System.Collections partial facade

FiveTimesTheFun opened the pull request:
Initial commit of System.Collections partial facade.  This partial
facade uses the new FindPartialFacadeContractReference and
FillPartialFacade targets to create a facade against mscorlib.  We
compile against the implementation assemblies mscorlib.dll (from the
Microsoft.DotNet.CoreCLR package), System.Diagnostics.Debug.dll,
and System.Runtime.dll.

Tests continue to compile against the System.Collections nuget package
until we have the ability to reference the contract for compilation
but copy the correct implementation assembly at test time.

stephentoub commented:
> Tests continue to compile against the System.Collections nuget package until we have the ability to reference the contract for compilation but copy the correct implementation assembly at test time.

@mellinoe appears to have worked around that in System.Runtime.Handles, assuming I'm reading the comment correctly.  Could we do the same thing here?
https://github.com/dotnet/corefx/commit/a7acef765e77e0a1dd80e3d1e56f00b74402028a#diff-4cdbf1333a93ee83e7c15b05f49eb713R18

stephentoub commented on src/Common/src/System/Collections/Generic/LowLevelComparer.cs:
sealed?

stephentoub commented on src/System.Collections/src/System/Collections/Generic/Queue.cs:
```diff
@@ -0,0 +1,461 @@
+        private Object _syncRoot;
+
+        private const int _MinimumGrow = 4;
+        private const int _ShrinkThreshold = 32;
```
While it might be nice if this was used, it doesn't appear to be.

FiveTimesTheFun replied:
That looks accurate.  I checked the rest of the fields and they're used.  I've pushed a commit to remove _ShrinkThreshold.

stephentoub commented on src/System.Collections/src/System/Collections/BitArray.cs:
```diff
@@ -0,0 +1,544 @@
+        private int _version;
+        private Object _syncRoot;
+
+        private const int _ShrinkThreshold = 256;
```
Per style guidelines, constants aren't supposed to begin with underscores, right?

FiveTimesTheFun replied:
Yep, thanks Stephen.  Updated, and filed https://github.com/dotnet/codeformatter/issues/117 .

stephentoub commented on src/System.Collections/src/System/Collections/Generic/BitHelper.cs:
```diff
@@ -0,0 +1,153 @@
+    /// every bit, it would allow good perf improvements when the bit array is sparse.
+    /// </summary>
+    unsafe internal class BitHelper
+    {   // should not be serialized
```
What does this "should not be serialized" comment refer to?

FiveTimesTheFun replied:
I believe it's because the class has a local field that is a pointer to stack memory.

stephentoub commented on src/System.Collections/src/System/Collections/Generic/BitHelper.cs:
```diff
@@ -0,0 +1,153 @@
+                if (bitArrayIndex < m_length && bitArrayIndex >= 0)
+                {
+                    m_array[bitArrayIndex] |= (MarkedBitFlag << (bitPosition % IntSize));
+                }
```
In both this and the subsequent function, the only difference between the if and else block appears to be whether the result is stored into m_arrayPtr or m_array.  You could avoid duplicated code by just having the conditional around that piece rather than the whole thing.

FiveTimesTheFun replied:
Thanks, updated.

stephentoub commented on src/System.Collections/src/System/Collections/Generic/Comparer.cs:
```diff
@@ -0,0 +1,94 @@
+{
+    public abstract class Comparer<T> : IComparer, IComparer<T>
+    {
+        protected Comparer()
```
This shouldn't be necessary. Is the ctor here just to be explicit?

FiveTimesTheFun replied:
Comparer is type forwarded to mscorlib, I've updated my initial commit to remove this file.

stephentoub commented on src/System.Collections/src/System/Collections/Generic/Comparer.cs:
```diff
@@ -0,0 +1,94 @@
+            get
+            {
+                if (_default == null)
+                    _default = CreateComparer();
```
Probably something to be done as a separate commit, but see https://github.com/dotnet/corefx/issues/780

FiveTimesTheFun replied:
Comparer is type forwarded to mscorlib, I've updated my initial commit to remove this file.

stephentoub commented on src/Common/src/System/Collections/Generic/LowLevelComparer.cs:
```diff
@@ -0,0 +1,104 @@
+
+namespace System.Collections.Generic
+{
+    internal sealed class LowLevelComparer<T> : IComparer<T>
```
Why is this type necessary?  Why is just Comparer<T> (also in this assembly) insufficient?

weshaggard replied:
I have a similar question.

FiveTimesTheFun replied:
This doesn't appear to be used anymore, I've removed it.

stephentoub commented on src/System.Collections/src/System/Collections/Generic/Comparer.cs:
```diff
@@ -0,0 +1,94 @@
+
+        private static Comparer<T> CreateComparer()
+        {
+            // NUTC compiler optimization see comments in EqualityComparer.cs
```
See my comment on the same code in ```LowLevelComparer<T>```.

stephentoub commented on src/System.Collections/src/System/Collections/Generic/Comparer.cs:
```diff
@@ -0,0 +1,94 @@
+    }
+
+
+    internal class DefaultComparer<T> : Comparer<T>
```
Just for hygiene reasons, it'd be nice if all of these derived types could be sealed.

FiveTimesTheFun replied:
Thanks, updated.

stephentoub commented on src/System.Collections/src/System/Collections/Generic/Dictionary.cs:
```diff
@@ -0,0 +1,1265 @@
+        private ValueCollection values;
+        private Object _syncRoot;
+
+        // constants for serialization
```
Are these constants used anywhere?

FiveTimesTheFun replied:
Dictionary is type forwarded to mscorlib, I've updated my initial commit to remove this file.

stephentoub commented on src/System.Collections/src/System/Collections/Generic/Dictionary.cs:
```diff
@@ -0,0 +1,1265 @@
+        {
+            if (count > 0)
+            {
+                for (int i = 0; i < buckets.Length; i++) buckets[i] = -1;
```
I forget now... did we add anything to the style guidelines around one line loops like this?

FiveTimesTheFun replied:
I don't see anything on the Wiki:
https://github.com/dotnet/corefx/wiki/Coding-style

Willing to start an issue to discuss, though.  Would personally like to keep constructs like
if (expr) return val1;
else return val2;
but will update this one for better readability.

stephentoub commented on src/System.Collections/src/System/Collections/Generic/Dictionary.cs:
```diff
@@ -0,0 +1,1265 @@
+            int hashCode = comparer.GetHashCode(key) & 0x7FFFFFFF;
+            int targetBucket = hashCode % buckets.Length;
+
+#if FEATURE_RANDOMIZED_STRING_HASHING
```
Is this compilation constant still needed?  We don't appear to be setting it; should we?

FiveTimesTheFun replied:
Thanks Stephen - this feature is enabled for coreclr (and set for mscorlib, https://github.com/dotnet/coreclr/blob/524f5463406df065474b342086e6dfe38ddfbda3/clr.coreclr.props#L78 ) and should be enabled here.

FiveTimesTheFun replied:
It looks like it should also be enabled for https://github.com/dotnet/corefx/blob/7116584186f8f3a886616aaf8cb5d4a982c60e27/src/System.Collections.NonGeneric/src/System/Collections/Hashtable.cs#L821 .

It looks like our corefx common msbuild files only define "DEBUG" and "TRACE", while our buildtools msbuild files only define "DEBUGRESOURCES" and "SIGNED", and individual libraries have so far defined 'feature' constants like this one themselves.  I'll update this PR and add a new PR for System.Collections.NonGeneric.

FiveTimesTheFun replied:
We type forward to mscorlib for Dictionary here, but the same comment applies to HashSet.

FiveTimesTheFun replied:
Filed https://github.com/dotnet/corefx/issues/1534

stephentoub commented on src/System.Collections/src/System/Collections/Generic/EqualityComparer.cs:
```diff
@@ -0,0 +1,336 @@
+                {
+                    object comparer;
+
+                    // NUTC compiler is able to static evalulate the conditions and only put the necessary branches in finally binary code, 
```
The same is true of the JIT; at least it was in .NET 4.5.  Hopefully that hasn't changed with RyuJIT.

FiveTimesTheFun replied:
EqualityComparer is type forwarded to mscorlib, I've updated my initial commit to remove this file.

stephentoub commented on src/System.Collections/src/System/Collections/Generic/EqualityComparer.cs:
```diff
@@ -0,0 +1,336 @@
+    }
+
+    //
+    // ProjectN compatiblity notes:
```
Is this comment still relevant?  And this should be ".NET Native" rather than "ProjectN", right?

FiveTimesTheFun replied:
EqualityComparer is type forwarded to mscorlib, I've updated my initial commit to remove this file.

FiveTimesTheFun commented:
Thanks Stephen - I updated the test project to copy the product assembly to the output directory through a project reference and updated the initial commit.  I'm going through the rest of your comments now.

stephentoub commented on src/System.Collections/src/System/Collections/Generic/LinkedList.cs:
```diff
@@ -0,0 +1,606 @@
+        private Object _syncRoot;
+
+        // names for serialization
+        const String VersionName = "Version";
```
Same question about the serialization constants.

FiveTimesTheFun replied:
I believe these were used in GetObjectData and OnDeserialization, support for which was removed from this class.  I'll remove the fields.

FiveTimesTheFun replied:
Updated.

stephentoub commented on src/System.Collections/src/System/Collections/Generic/LinkedList.cs:
```diff
@@ -0,0 +1,606 @@
+    }
+
+    // Note following class is not serializable since we customized the serialization of LinkedList. 
+    [System.Runtime.InteropServices.ComVisible(false)]
```
Are such ComVisible attributes still necessary?

FiveTimesTheFun replied:
There aren't that many others - I was under the impression this was in the same bucket as security transparency attributes - don't add but leave.

FiveTimesTheFun replied:
Apparently the guidance has changed (for security transparency attributes, too).
Updated to remove all the ComVisibleAttributes and Security*Attributes.

stephentoub commented on src/System.Collections/src/System/Collections/Generic/List.cs:
```diff
@@ -0,0 +1,1381 @@
+        private int _version;
+        private Object _syncRoot;
+
+        static readonly T[] _emptyArray = new T[0];
```
Style guideline naming: this should be s_emptyArray.  It should also be explicitly marked as private.

stephentoub replied:
Consider using Array.Empty<T>() instead of these _emptyArray fields in these various types.

FiveTimesTheFun replied:
Thanks - using Array.Empty<T>() in Queue and Stack.

stephentoub commented on src/System.Collections/src/System/Collections/Generic/List.cs:
```diff
@@ -0,0 +1,1381 @@
+    [DebuggerDisplay("Count = {Count}")]
+    public class List<T> : IList<T>, System.Collections.IList, IReadOnlyList<T>
+    {
+        private const int _defaultCapacity = 4;
```
Style guideline naming: Constants should be PascalCased, without a leading underscore.

FiveTimesTheFun replied:
Thanks Stephen - updated.

FiveTimesTheFun commented on src/Common/src/System/Collections/Generic/LowLevelComparer.cs:
Thanks Stephen - I updated to make all internal classes sealed where appropriate.

stephentoub commented on src/System.Collections/src/System/Collections/Generic/Queue.cs:
```diff
@@ -0,0 +1,461 @@
+        private const int _ShrinkThreshold = 32;
+        private const int _GrowFactor = 200;  // double each time
+        private const int _DefaultCapacity = 4;
+        static T[] _emptyArray = new T[0];
```
Same comments about static field and constant names... I'll stop commenting on these :)

FiveTimesTheFun replied:
Thanks Stephen.  Code formatter fixed all the statics, and I manually fixed all the constants.

FiveTimesTheFun replied:
Updated.

stephentoub commented on src/System.Collections/src/System/Collections/Generic/QueueDebugView.cs:
```diff
@@ -0,0 +1,31 @@
+
+namespace System.Collections.Generic
+{
+    internal sealed class QueueDebugView<T>
```
Since most of these debug views just do ToArray, could we collapse these into views based on the interfaces?

FiveTimesTheFun replied:
That would be ICollection<T>, right?  We already have a separate debug view for that, ICollectionDebugView<T>, which makes a copy of the contents.

stephentoub replied:
If the end result is the same, can we just use that then, rather than having a debug view specific to Queue?

FiveTimesTheFun replied:
Sure - I've just done this for HashSet and SortedSet.  I didn't do this for Queue and Stack, since they don't implement the generic ICollection<> interface.  The extension method for ToArray(this IEnumerable<>) is up in System.Linq.

stephentoub commented on src/System.Collections/src/System/Collections/Generic/SortedSet.cs:
```diff
@@ -0,0 +1,2552 @@
+        int version;
+        private Object _syncRoot;
+
+        private const String ComparerName = "Comparer";
```
Same question about serialization constants.  Will stop commenting on these.

FiveTimesTheFun replied:
Thank you - updated.

stephentoub commented on src/System.Collections/src/System/Collections/Generic/SortedSet.cs:
```diff
@@ -0,0 +1,2552 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+// #define USING_HASH_SET
```
Can the #ifs for USING_HASH_SET be deleted, or should this be added to the .csproj?

FiveTimesTheFun replied:
Unfortunately my history is incomplete for when that change was entered in 2008.  All of the methods that use that define were introduced at the same time as the define in my history.  It's currently unset on ProjectK and desktop.  I'll remove this.

FiveTimesTheFun replied:
Updated.

stephentoub commented on src/System.Collections/src/project.json:
```diff
@@ -0,0 +1,24 @@
+{
+  "dependencies": {
+//    "System.Collections": "4.0.10-beta-22809",
```
What are all of these '//'s?

FiveTimesTheFun replied:
Pardon, I've removed those.

stephentoub commented on src/System.Collections/src/project.json:
```diff
@@ -0,0 +1,24 @@
+//    "System.Threading.Tasks": "4.0.10-beta-22809",
+//    "System.Threading.ThreadPool": "4.0.10-beta-22809"
+  }
+}
```
Nit: missing newline

FiveTimesTheFun replied:
Thanks - updated.

FiveTimesTheFun commented:
Thanks Stephen and Wes - I believe I've addressed all the comments and rebased this on top of master.

FiveTimesTheFun commented:
Thanks @stephentoub and @weshaggard  - I believe I've addressed all the comments and rebased this on top of master.

mellinoe commented:
You'll want to add the line to the test project file, as Stephen did here:

https://github.com/dotnet/corefx/pull/1541

Other than that, this looks good to me.

krwq commented on src/Common/src/System/Collections/HashHelpers.cs:
```diff
@@ -0,0 +1,107 @@
+            1103, 1327, 1597, 1931, 2333, 2801, 3371, 4049, 4861, 5839, 7013, 8419, 10103, 12143, 14591,
+            17519, 21023, 25229, 30293, 36353, 43627, 52361, 62851, 75431, 90523, 108631, 130363, 156437,
+            187751, 225307, 270371, 324449, 389357, 467237, 560689, 672827, 807403, 968897, 1162687, 1395263,
+            1674319, 2009191, 2411033, 2893249, 3471899, 4166287, 4999559, 5999471, 7199369};
```
Nit: it feels like there is a spacebar before } missing. Could be formatter bug

nydehi replied:
isnt  5 a prime as well?? or is that left out for a reason?

FiveTimesTheFun replied:
@krwq - updated
@nydehi  - this list of primes is not exhaustive - it's missing many more than 5.  This list is used to choose a hash table size for HashSet.  By having fewer primes in this list we speed up the implementation of GetPrime(int) with a tradeoff of possibly returning a higher prime than would be necessary.

FiveTimesTheFun commented:
@krwq - The latest guidance from @weshaggard is to remove stale ComVisible attributes, SecurityCritical/SafeCritical/Transparent attributes because those two features are not supported on .NET Core, and while leaving the attributes while porting doesn't harm anything, they become stale and inaccurate or incomplete over time.

FiveTimesTheFun commented:
Thanks @mellinoe  - updated.

FiveTimesTheFun commented:
@dotnet-bot test this please.
